zebra-chain/src/versioned_sig.rs

```diff
@@ -0,0 +1,257 @@
+//! Versioned signature types for transaction V6+.
+//!
+//! ZIP 246 introduces sighash versioning to allow future upgrades to signature algorithms.
+//! Each signature is prefixed with sighash metadata (version + optional associated data).
+
+use std::io;
+
+use crate::serialization::{
+    CompactSizeMessage, SerializationError, TrustedPreallocate, ZcashDeserialize,
+    ZcashDeserializeInto, ZcashSerialize,
+};
+
+/// Sighash version byte.
+///
+/// Each transaction version defines its own set of sighash versions.
+/// For V6 transactions, only version 0 is currently defined.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct SighashVersion(pub u8);
+
+impl SighashVersion {
+    /// Sighash version 0 (default for V6 transactions).
+    pub const V0: Self = SighashVersion(0);
+}
+
+/// Sighash metadata for V6+ transaction signatures.
+///
+/// Serialization format: `CompactSize(len) || version || associated_data`
+///
+/// For V6 transactions with version 0, `associated_data` is always empty.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SighashInfo {
+    /// Sighash algorithm version.
+    pub version: SighashVersion,
+
+    /// Optional data specific to this sighash version.
+    pub associated_data: Vec<u8>,
+}
+
+impl SighashInfo {
+    /// Create V0 sighash info for transaction V6 (empty `associated_data`).
+    pub fn for_tx_v6() -> Self {
+        SighashInfo {
+            version: SighashVersion::V0,
+            associated_data: Vec::new(),
+        }
+    }
+
+    /// Validate this is V0 format for transaction V6.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - Version is not 0
+    /// - `associated_data` is not empty
+    pub fn as_tx_v6(&self) -> Result<(), SerializationError> {
+        if self.version != SighashVersion::V0 {
+            return Err(SerializationError::Parse(
+                "V6 transactions only support sighash version 0",
+            ));
+        }
+        if !self.associated_data.is_empty() {
+            return Err(SerializationError::Parse(
+                "V6 sighash v0 must have empty associated_data",
+            ));
```

Comment:
instead of these two manual checks, it would be more elegant to generate an actual ` SighashInfo::v0()` and compare.